### ISSUE TITLE
module version 0.15.1 breaks support for Ubuntu 14.04 LTS

### ISSUE THREAD

**Issue by anderssynstad** [elastic/puppet-elasticsearch#752]
* Module version: 0.15.1
* Puppet version: 3.4.3
* OS and version: Ubuntu 14.04 LTS

With commit  ddb00d699567ea9971afbd25350bdb5218187b64 and the use of the "new" digest_algorithm method in puppet, one bumped the puppet requirement to >=3.8.0, as mentioned in commit  903e8129b4c6207d23f53e451242b4ac8a3534eb.

Since Ubuntu 14.04 LTS comes with puppet 3.4.3, this breaks support for Ubuntu LTS releases older than 16.04 (unless using puppet from puppetlabs directly).

metadata.json still says:
>     {
>       "operatingsystem": "Ubuntu",
>       "operatingsystemrelease": [
>         "10.04",
>         "12.04",
>         "14.04",
>         "16.04"
>       ]
>     },
> 

and README.md still says:
> Limitations
> 
> This module has been built on and tested against Puppet 3.2 and higher.
> 
> The module has been tested on:
> 
> Debian 6/7/8
> CentOS 6/7
> OracleLinux 6/7
> Ubuntu 12.04, 14.04
> OpenSuSE 13.x
> SLES 12

So at least README.md and metadata.json should be updated to reflect this change, but is commit  ddb00d699567ea9971afbd25350bdb5218187b64 worth breaking support for Ubuntu 14.04 LTS for? (have not checked if there are other changes requiring puppet >=3.8.0)

**Comment by tylerjl** [comment]
Thanks for bringing this up, I think it's important to lay out what the perspective of this module is as it pertains to Puppet versions. A few points worth mentioning:
- 14.04 _is_ still supported. It's tested regularly in CI (albeit with the requirement that it does require the minimum required version in the module metadata)
- As you mentioned, some versions like 14.04 will rely on the official Puppet package repositories for newer versions of Puppet. It's not ideal, but it's a pretty huge stretch to try and support both an older distribution _and_ a very old version of Puppet.
- I think it's a pretty fair expectation for the module to follow Puppet Labs' support of Puppet versions. Note that the latest in the 3.x series is slated to be dropped from maintenance and security updates - not just updates, but legacy support entirely - [in 2017](https://groups.google.com/forum/#!msg/puppet-dev/jkcCNSk9w4w/3ltC-wibDAAJ). _Even with that announcement of deprecation from the creators of the language_, I'm going to try to keep >= 3.8 functional as long as possible. To try and extend this back to 3.4 is, bluntly, a little ridiculous given that even 3.8 is about to be completely unsupported by Puppet Labs.
- You noted one commit that necessitates Puppet version >= 3.8, I know of at least one other instance off-hand in 7f086aee4c9 (which uses `file()` in a way that I think changed in 3.7?)
- You've got a totally valid point in that I probably need to remove Ubuntu 10.04 from the supported list as that isn't tested at all and don't know whether it works with the current module at all.

**tl;dr** 14.04 is still supported, it just needs the right version of Puppet (which is available for 14.04), and the main cutoff from the module's perspective is Puppet versions.

I do sympathize with users on older distributions as it isn't as straightforward as it could be; but it's certainly possible to do with the upstream repositories, and we've got to move the module forward as time goes on.

**Comment by tylerjl** [comment]
I'm going to close this out, but if anyone has lingering concerns, please do bring them up for discussion so we can come to a solution. 👍
